File: surface/widgets/swiftui/components/widgets/concepts/llm-agent/TraceTree.swift
import SwiftUI

// MARK: - Types

struct TraceSpan: Identifiable {
    let id: String
    let type: String
    let label: String
    let duration: Int
    var tokens: Int?
    let status: String
    var children: [TraceSpan]?
}

// MARK: - State Machine

enum TraceTreeState { case idle, spanSelected }
enum TraceTreeEvent {
    case selectSpan(id: String), deselect, expand(id: String), collapse(id: String), filter(spanType: String)
}

func traceTreeReduce(state: TraceTreeState, event: TraceTreeEvent) -> TraceTreeState {
    switch state {
    case .idle:
        if case .selectSpan = event { return .spanSelected }
        return state
    case .spanSelected:
        switch event {
        case .deselect: return .idle
        case .selectSpan: return .spanSelected
        default: return state
        }
    }
}

private let spanTypeLabels: [String: String] = ["llm": "LLM", "tool": "Tool", "chain": "Chain", "agent": "Agent"]
private let spanTypeIcons: [String: String] = ["llm": "\u{1F9E0}", "tool": "\u{2699}", "chain": "\u{1F517}", "agent": "\u{1F916}"]
private let spanStatusIcons: [String: String] = ["success": "\u{2713}", "running": "\u{25CB}", "error": "\u{2717}", "pending": "\u{2022}"]

private func findSpan(_ spans: [TraceSpan], id: String) -> TraceSpan? {
    for s in spans {
        if s.id == id { return s }
        if let c = s.children, let f = findSpan(c, id: id) { return f }
    }
    return nil
}

// MARK: - View

struct TraceTreeView: View {
    let spans: [TraceSpan]
    let rootLabel: String
    var totalDuration: Int?
    var totalTokens: Int?
    var onSelectSpan: ((String?) -> Void)?
    var showMetrics: Bool = true

    @State private var widgetState: TraceTreeState = .idle
    @State private var selectedId: String? = nil
    @State private var expandedIds: Set<String> = []
    @State private var visibleTypes: Set<String> = ["llm", "tool", "chain", "agent"]

    private var selectedSpan: TraceSpan? {
        guard let sid = selectedId else { return nil }
        return findSpan(spans, id: sid)
    }

    private var availableTypes: [String] {
        var types: Set<String> = []
        func walk(_ nodes: [TraceSpan]) { for s in nodes { types.insert(s.type); if let c = s.children { walk(c) } } }
        walk(spans); return Array(types).sorted()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Header
            HStack {
                Text(rootLabel).font(.headline)
                Spacer()
                if showMetrics, let d = totalDuration { Text("\(d)ms").font(.caption).foregroundColor(.secondary) }
                if showMetrics, let t = totalTokens { Text("\(t) tokens").font(.caption).foregroundColor(.secondary) }
            }.padding(.horizontal, 12).padding(.vertical, 8)

            // Filter bar
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 4) {
                    ForEach(availableTypes, id: \.self) { t in
                        Button("\(spanTypeLabels[t] ?? t)") {
                            if visibleTypes.contains(t) { visibleTypes.remove(t) } else { visibleTypes.insert(t) }
                        }
                        .font(.system(size: 12))
                        .fontWeight(visibleTypes.contains(t) ? .semibold : .regular)
                        .padding(.horizontal, 8).padding(.vertical, 2)
                        .background(visibleTypes.contains(t) ? Color.blue.opacity(0.15) : Color.clear)
                        .cornerRadius(4).buttonStyle(.plain)
                    }
                }
            }.padding(.horizontal, 12).padding(.bottom, 6)

            Divider()

            HStack(alignment: .top, spacing: 0) {
                ScrollView {
                    LazyVStack(alignment: .leading, spacing: 0) {
                        ForEach(spans.filter { visibleTypes.contains($0.type) }) { span in
                            spanNode(span, depth: 0)
                        }
                    }
                }

                if widgetState == .spanSelected, let span = selectedSpan {
                    Divider()
                    detailPanel(span).frame(width: 240)
                }
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Execution trace")
    }

    @ViewBuilder
    private func spanNode(_ span: TraceSpan, depth: Int) -> some View {
        let isExpanded = expandedIds.contains(span.id)
        let isSelected = selectedId == span.id
        let hasChildren = !(span.children ?? []).isEmpty
        let visibleChildren = (span.children ?? []).filter { visibleTypes.contains($0.type) }

        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 6) {
                if hasChildren {
                    Button(action: { toggleExpand(span.id) }) {
                        Text(isExpanded ? "\u{25BC}" : "\u{25B6}").font(.system(size: 10))
                    }.buttonStyle(.plain)
                }
                Text(spanTypeIcons[span.type] ?? "\u{25CF}").font(.system(size: 12))
                Text(span.label).lineLimit(1)
                Text("\(span.duration)ms").font(.caption).foregroundColor(.secondary)
                if showMetrics, let t = span.tokens { Text("\(t) tok").font(.caption2).foregroundColor(.secondary) }
                Spacer()
                Text(spanStatusIcons[span.status] ?? "\u{2022}")
                    .foregroundColor(span.status == "error" ? .red : .primary)
            }
            .padding(.leading, CGFloat(depth * 16))
            .padding(.horizontal, 8).padding(.vertical, 5)
            .background(isSelected ? Color.blue.opacity(0.1) : Color.clear)
            .contentShape(Rectangle())
            .onTapGesture { selectSpan(span.id) }
            .accessibilityLabel("\(span.type): \(span.label) (\(span.duration)ms)")

            if hasChildren && isExpanded {
                ForEach(visibleChildren) { child in spanNode(child, depth: depth + 1) }
            }
        }
    }

    @ViewBuilder
    private func detailPanel(_ span: TraceSpan) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Text("\(spanTypeIcons[span.type] ?? "") \(spanTypeLabels[span.type] ?? span.type)").font(.caption).foregroundColor(.secondary)
                Spacer()
                Button(action: { deselectSpan() }) { Text("\u{2715}") }.buttonStyle(.plain)
            }
            Group {
                labelValue("Label", span.label)
                labelValue("Status", "\(spanStatusIcons[span.status] ?? "") \(span.status)")
                labelValue("Duration", "\(span.duration)ms")
                if let t = span.tokens { labelValue("Tokens", "\(t)") }
                if let c = span.children, !c.isEmpty { labelValue("Children", "\(c.count) spans") }
            }
            Spacer()
        }.padding()
    }

    @ViewBuilder
    private func labelValue(_ label: String, _ value: String) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label).font(.caption).foregroundColor(.secondary)
            Text(value).font(.system(size: 13))
        }
    }

    private func selectSpan(_ id: String) {
        if selectedId == id { deselectSpan(); return }
        selectedId = id
        widgetState = traceTreeReduce(state: widgetState, event: .selectSpan(id: id))
        onSelectSpan?(id)
    }

    private func deselectSpan() {
        selectedId = nil
        widgetState = traceTreeReduce(state: widgetState, event: .deselect)
        onSelectSpan?(nil)
    }

    private func toggleExpand(_ id: String) {
        if expandedIds.contains(id) { expandedIds.remove(id) } else { expandedIds.insert(id) }
    }
}

File: surface/widgets/nativescript/components/widgets/concepts/llm-agent/TraceTree.ts
import { StackLayout, Label, Button, ScrollView, FlexboxLayout } from '@nativescript/core';

/* ---------------------------------------------------------------------------
 * TraceTree state machine
 * States: idle (initial), spanSelected (with detail panel)
 * ------------------------------------------------------------------------- */

export type TraceTreeState = 'idle' | 'spanSelected';
export type TraceTreeEvent =
  | { type: 'SELECT_SPAN'; id: string }
  | { type: 'DESELECT' }
  | { type: 'EXPAND'; id: string }
  | { type: 'COLLAPSE'; id: string }
  | { type: 'FILTER'; spanType: string };

export function traceTreeReducer(state: TraceTreeState, event: TraceTreeEvent): TraceTreeState {
  switch (state) {
    case 'idle':
      if (event.type === 'SELECT_SPAN') return 'spanSelected';
      return state;
    case 'spanSelected':
      if (event.type === 'DESELECT') return 'idle';
      if (event.type === 'SELECT_SPAN') return 'spanSelected';
      return state;
    default:
      return state;
  }
}

/* ---------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

export interface TraceSpan {
  id: string;
  type: string;
  label: string;
  duration: number;
  tokens?: number;
  status: string;
  children?: TraceSpan[];
}

export interface TraceTreeProps {
  spans: TraceSpan[];
  rootLabel: string;
  totalDuration?: number;
  totalTokens?: number;
  selectedSpanId?: string;
  onSelectSpan?: (id: string | undefined) => void;
  expandedIds?: string[];
  visibleTypes?: string[];
  showMetrics?: boolean;
}

/* ---------------------------------------------------------------------------
 * Helpers
 * ------------------------------------------------------------------------- */

const SPAN_TYPE_LABELS: Record<string, string> = {
  llm: 'LLM', tool: 'Tool', chain: 'Chain', agent: 'Agent',
};

const SPAN_TYPE_ICONS: Record<string, string> = {
  llm: '\uD83E\uDDE0', tool: '\u2699', chain: '\uD83D\uDD17', agent: '\uD83E\uDD16',
};

const STATUS_ICONS: Record<string, string> = {
  success: '\u2713', running: '\u25CB', error: '\u2717', pending: '\u2022',
};

function findSpan(spans: TraceSpan[], id: string): TraceSpan | undefined {
  for (const span of spans) {
    if (span.id === id) return span;
    if (span.children?.length) {
      const found = findSpan(span.children, id);
      if (found) return found;
    }
  }
  return undefined;
}

function collectTypes(spans: TraceSpan[]): string[] {
  const types = new Set<string>();
  function walk(nodes: TraceSpan[]) {
    for (const s of nodes) {
      types.add(s.type);
      if (s.children?.length) walk(s.children);
    }
  }
  walk(spans);
  return Array.from(types);
}

/* ---------------------------------------------------------------------------
 * Widget
 * ------------------------------------------------------------------------- */

export function createTraceTree(props: TraceTreeProps): { view: StackLayout; dispose: () => void } {
  const {
    spans,
    rootLabel,
    totalDuration,
    totalTokens,
    selectedSpanId: initialSelectedId,
    onSelectSpan,
    expandedIds: controlledExpandedIds,
    visibleTypes: controlledVisibleTypes,
    showMetrics = true,
  } = props;

  let selectedId: string | undefined = initialSelectedId;
  let expandedSet = new Set<string>(controlledExpandedIds ?? []);
  let visibleSet = new Set<string>(controlledVisibleTypes ?? ['llm', 'tool', 'chain', 'agent']);
  const disposers: (() => void)[] = [];

  const root = new StackLayout();
  root.className = 'trace-tree';
  root.automationText = 'Execution trace';

  // Header
  const header = new FlexboxLayout();
  header.className = 'trace-tree-header';
  header.flexDirection = 'row' as any;
  header.alignItems = 'center' as any;

  const rootLabelView = new Label();
  rootLabelView.className = 'trace-tree-root-label';
  rootLabelView.text = rootLabel;
  header.addChild(rootLabelView);

  if (showMetrics && totalDuration != null) {
    const durationLabel = new Label();
    durationLabel.className = 'trace-tree-total-duration';
    durationLabel.text = `${totalDuration}ms`;
    header.addChild(durationLabel);
  }

  if (showMetrics && totalTokens != null) {
    const tokensLabel = new Label();
    tokensLabel.className = 'trace-tree-total-tokens';
    tokensLabel.text = `${totalTokens} tokens`;
    header.addChild(tokensLabel);
  }

  root.addChild(header);

  // Filter bar
  const filterBar = new FlexboxLayout();
  filterBar.className = 'trace-tree-filter-bar';
  filterBar.flexDirection = 'row' as any;

  const availableTypes = collectTypes(spans);
  for (const spanType of availableTypes) {
    const btn = new Button();
    btn.className = visibleSet.has(spanType) ? 'trace-tree-filter-active' : 'trace-tree-filter';
    btn.text = SPAN_TYPE_LABELS[spanType] ?? spanType;
    const handler = () => {
      if (visibleSet.has(spanType)) visibleSet.delete(spanType);
      else visibleSet.add(spanType);
      rebuildTree();
    };
    btn.on('tap', handler);
    disposers.push(() => btn.off('tap', handler));
    filterBar.addChild(btn);
  }
  root.addChild(filterBar);

  // Tree area
  const scrollView = new ScrollView();
  const treeContainer = new StackLayout();
  treeContainer.className = 'trace-tree-spans';
  scrollView.content = treeContainer;
  root.addChild(scrollView);

  // Detail panel
  const detailPanel = new StackLayout();
  detailPanel.className = 'trace-tree-detail-panel';
  detailPanel.visibility = 'collapse' as any;
  root.addChild(detailPanel);

  function renderSpan(span: TraceSpan, depth: number, container: StackLayout) {
    if (!visibleSet.has(span.type)) return;

    const hasChildren = !!(span.children?.length);
    const isExpanded = expandedSet.has(span.id);
    const isSelected = selectedId === span.id;

    const spanRow = new FlexboxLayout();
    spanRow.className = isSelected ? 'trace-tree-span-selected' : 'trace-tree-span';
    spanRow.flexDirection = 'row' as any;
    spanRow.alignItems = 'center' as any;
    spanRow.paddingLeft = depth * 16;
    spanRow.automationText = `${span.type}: ${span.label} (${span.duration}ms)`;

    if (hasChildren) {
      const expandLabel = new Label();
      expandLabel.className = 'trace-tree-expand';
      expandLabel.text = isExpanded ? '\u25BC' : '\u25B6';
      const expandHandler = () => {
        if (expandedSet.has(span.id)) expandedSet.delete(span.id);
        else expandedSet.add(span.id);
        rebuildTree();
      };
      expandLabel.on('tap', expandHandler);
      spanRow.addChild(expandLabel);
    }

    const iconLabel = new Label();
    iconLabel.className = 'trace-tree-span-icon';
    iconLabel.text = SPAN_TYPE_ICONS[span.type] ?? '\u25CF';
    spanRow.addChild(iconLabel);

    const labelView = new Label();
    labelView.className = 'trace-tree-span-label';
    labelView.text = span.label;
    spanRow.addChild(labelView);

    const durationView = new Label();
    durationView.className = 'trace-tree-span-duration';
    durationView.text = `${span.duration}ms`;
    spanRow.addChild(durationView);

    if (showMetrics && span.tokens != null) {
      const tokensView = new Label();
      tokensView.className = 'trace-tree-span-tokens';
      tokensView.text = `${span.tokens} tok`;
      spanRow.addChild(tokensView);
    }

    const statusView = new Label();
    statusView.className = 'trace-tree-span-status';
    statusView.text = STATUS_ICONS[span.status] ?? '\u2022';
    spanRow.addChild(statusView);

    const tapHandler = () => {
      selectedId = selectedId === span.id ? undefined : span.id;
      onSelectSpan?.(selectedId);
      rebuildTree();
      updateDetailPanel();
    };
    spanRow.on('tap', tapHandler);

    container.addChild(spanRow);

    if (hasChildren && isExpanded) {
      const visibleChildren = span.children!.filter((c) => visibleSet.has(c.type));
      for (const child of visibleChildren) {
        renderSpan(child, depth + 1, container);
      }
    }
  }

  function rebuildTree() {
    treeContainer.removeChildren();
    const visibleSpans = spans.filter((s) => visibleSet.has(s.type));
    for (const span of visibleSpans) {
      renderSpan(span, 0, treeContainer);
    }
    updateDetailPanel();
  }

  function updateDetailPanel() {
    detailPanel.removeChildren();
    if (!selectedId) {
      detailPanel.visibility = 'collapse' as any;
      return;
    }

    const span = findSpan(spans, selectedId);
    if (!span) {
      detailPanel.visibility = 'collapse' as any;
      return;
    }

    detailPanel.visibility = 'visible' as any;

    const detailHeader = new FlexboxLayout();
    detailHeader.flexDirection = 'row' as any;
    detailHeader.alignItems = 'center' as any;

    const typeLabel = new Label();
    typeLabel.text = `${SPAN_TYPE_ICONS[span.type] ?? '\u25CF'} ${SPAN_TYPE_LABELS[span.type] ?? span.type}`;
    detailHeader.addChild(typeLabel);

    const closeBtn = new Button();
    closeBtn.text = '\u2715';
    closeBtn.automationText = 'Close detail panel';
    const closeHandler = () => {
      selectedId = undefined;
      onSelectSpan?.(undefined);
      rebuildTree();
    };
    closeBtn.on('tap', closeHandler);
    detailHeader.addChild(closeBtn);
    detailPanel.addChild(detailHeader);

    const fields: [string, string][] = [
      ['Label', span.label],
      ['Status', `${STATUS_ICONS[span.status] ?? '\u2022'} ${span.status}`],
      ['Duration', `${span.duration}ms`],
    ];
    if (span.tokens != null) fields.push(['Tokens', String(span.tokens)]);
    if (span.children && span.children.length > 0) fields.push(['Children', `${span.children.length} spans`]);

    for (const [label, value] of fields) {
      const fieldRow = new FlexboxLayout();
      fieldRow.flexDirection = 'row' as any;

      const fieldLabel = new Label();
      fieldLabel.className = 'trace-tree-detail-label';
      fieldLabel.text = label;
      fieldRow.addChild(fieldLabel);

      const fieldValue = new Label();
      fieldValue.className = 'trace-tree-detail-value';
      fieldValue.text = value;
      fieldRow.addChild(fieldValue);

      detailPanel.addChild(fieldRow);
    }
  }

  rebuildTree();

  return {
    view: root,
    dispose() {
      disposers.forEach((d) => d());
    },
  };
}

export default createTraceTree;
